RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+   ysum[nclusy+1] = 0.f;
+   ypos[nclusy+1] = yhigh0+0.5f*ysize;
+   
+   //   float xbaryc = xq/qtot;
+   
+   // Quit if only one pixel in cluster
+   
+      if(npixel < 2 ) {
+//         LOGDEBUG("SiPixelTemplateReco2D") << "2D fit not possible with single pixel" << ENDL;
+         return 1;
+      }
+   
+   
+   //   if(jmax-jmin == 0 ) {
+   //      LOGDEBUG("SiPixelTemplateReco2D") << "2D fit not possible with single x-pixel" << ENDL;
+   //      return 1;
+   //   }
+   
+   // Quit if only one pixel in cluster
+   
+   //   if(imax-imin == 0) {
+   //      LOGDEBUG("SiPixelTemplateReco2D") << "2D fit not possible with single y-pixel" << ENDL;
+   //      return 1;
+   //   }
+   
+   // Next, calculate the error^2 [need to know which is the middle y pixel of the cluster]
+   
+   int ypixoff = T2HYP1 - (imin+imax)/2;
+   for(k=0; k<npixel; ++k) {
+      i = indexxy[1][k];
+      int ypixeff = ypixoff + i;
+      templ2D.xysigma2(pixel[k], ypixeff, sigma2[k]);
+   }
+   
+   // Next, find the half-height edges of the y-projection and identify any
+   // missing columns to remove from the fit
+   
+   int imisslow = -1, imisshigh = -1, jmisslow = -1, jmisshigh = -1;
+   float ylow = -1.f, yhigh = -1.f;
+   float hmaxpix = fracpix*templ2D.sxymax();
+   for(i=imin; i<=imax; ++i) {
+      if(ysum[i] > hmaxpix && ysum[i-1] < hmaxpix && ylow < 0.f) {
+         ylow = ypos[i-1] + (ypos[i]-ypos[i-1])*(hmaxpix-ysum[i-1])/(ysum[i]-ysum[i-1]);
+      }
+      if(ysum[i] < q50) {
+         if(imisslow < 0) imisslow = i;
+         imisshigh = i;
+      }
+      if(ysum[i] > hmaxpix && ysum[i+1] < hmaxpix) {
+         yhigh = ypos[i] + (ypos[i+1]-ypos[i])*(ysum[i]-hmaxpix)/(ysum[i]-ysum[i+1]);
+      }
+   }
+   if(ylow < 0.f || yhigh < 0.f) {
+      ylow = ylow0;
+      yhigh = yhigh0;
+   }
+   
+   float templeny = templ2D.clsleny();
+   deltay = templeny - (yhigh - ylow)/ysize;
+   
+   // Calculate the corrected distance from the ylow and yhigh to the cluster center
+   
+//   float delycor = templ2D.delyavg();
+//   float halfy = 0.5f*(templeny-delycor)*ysize;
+   
+//   printf("templeny = %f, deltay = %f, delycor = %f, halfy = %f \n", templeny, deltay, delycor, halfy);
+   
+   //  x0 and y0 are best guess seeds for the fit
+   
+   //   float x0 = 0.5f*(xlow0 + xhigh0) + 0.5f*xsize;
+   float x0 = 0.5f*(xlow0 + xhigh0) - templ2D.lorxdrift();
+   float y0 = 0.5f*(ylow + yhigh) - templ2D.lorydrift();
+//   float y1 = yhigh - halfy - templ2D.lorydrift();
+//   printf("y0 = %f, y1 = %f \n", y0, y1);
+   //   float y0 = 0.5f*(ylow + yhigh);
+   
+   // If there are missing edge columns, set up missing column flags and number
+   // of minimization passes
+   
+   int npass = 1;
+   
+   switch(edgeflagy) {
+      case 0:
+         break;
+      case 1:
+//         y0 = yhigh - halfy - templ2D.lorydrift();
+         imisshigh = imin-1;
+         imisslow = -1;
+         break;
+      case 2:
+//         y0 = ylow + halfy - templ2D.lorydrift();

```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.